feat(encoding): switching to PyAV for ffmpeg related tasks

CarolinePascal opened the pull request:
## What this does

This PR leverages [PyAV](https://pyav.org/docs/stable/) `ffmpeg` API to handle all tasks relying on `ffmpeg` or `ffprobe` CLI commands (_i.e._ encoding and video/audio stream information gathering).

## How it was tested

Tested by recording a dummy dataset. The output information in `meta/info.json` should remain unchanged in comparison to the old `ffmpeg` CLI version.

## How to checkout & try? (for the reviewer)

CarolinePascal commented:
Waiting for your approval @aliberts 🫡

imstevenpmwork commented on lerobot/common/datasets/video_utils.py:
```diff
@@ -252,51 +251,68 @@ def encode_video_frames(
+            f"Video file already exists at {video_path}. Use `overwrite=True` to overwrite it."
+        )
+
     video_path.parent.mkdir(parents=True, exist_ok=True)
```
Instead of manually checking if `video_path` exists, you could do:

```python
video_path.parent.mkdir(parents=True, exist_ok=overwrite)
```

As `exist_ok` is essentially taking care of doing this

imstevenpmwork commented on lerobot/common/datasets/video_utils.py:
```diff
@@ -252,51 +251,68 @@ def encode_video_frames(
+        logging.getLogger("libav").setLevel(log_level)
+
+    # Create and open output file (overwrite by default)
+    with av.open(str(video_path), "w", format="mp4") as output:
```
Would it be helpful to pass the format as an input argument rather than hard-coding it?

CarolinePascal replied:
For now, we always save the output file as a MP4 file and I'm almost sure it won't change anytime soon.
It seems like adding it as an argument will more likely generate bugs :/

CarolinePascal replied:
[Update] I got why you were concerned, I think the format argument is a debug artefact as the format is inferred from the provided path !

CarolinePascal replied:
[d4782cb](https://github.com/huggingface/lerobot/pull/983/commits/d4782cbcce35d71f209febf137061c8164e586fe)

imstevenpmwork commented on lerobot/common/datasets/video_utils.py:
```diff
@@ -252,51 +251,68 @@ def encode_video_frames(
+        output_stream = output.add_stream(vcodec, fps, options=video_options)
+
+        # Loop through input frames and encode them
+        for input in input_list:
```
Commonly, it is wise to avoid [redefining built-in keywords](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/redefined-builtin.html)

imstevenpmwork commented on lerobot/common/datasets/video_utils.py:
```diff
@@ -252,51 +251,68 @@ def encode_video_frames(
+
+        # Loop through input frames and encode them
+        for input in input_list:
+            input_image = Image.open(input).convert("RGB")
```
Is the explicit conversion needed? What if my frames are not RGB ([e.g. BGR](https://github.com/huggingface/lerobot/blob/b43ece89340e7d250574ae7f5aaed5e8389114bd/lerobot/common/robot_devices/cameras/opencv.py#L391))? It seems anyways that `ImageFile` is a subclass of `Image`

CarolinePascal replied:
I was advised to put this convert here to eventually get rid of any 4th alpha channel which is not handled by PyAV. I checked if there was a way to remove this channel without changing the pixel format (RGB/BGR), but the only other way is to convert the fame into a numpy array and delete the 4th channel - not efficient.
Apparently, PIL is clever enough to convert the input frame properly regardless of the initial pixel format, and it's always used in the examples I've seen.

I would advise not to try to remove it.

imstevenpmwork commented on lerobot/common/datasets/video_utils.py:
```diff
@@ -252,51 +251,68 @@ def encode_video_frames(
     video_path = Path(video_path)
     imgs_dir = Path(imgs_dir)
+
+    if video_path.exists() and not overwrite:
```
Instead of manually checking if `video_path` exists, you could do:

```python
video_path.parent.mkdir(parents=True, exist_ok=overwrite)
```

As `exist_ok` is essentially taking care of doing this

CarolinePascal replied:
Fixed in [d4782cb](https://github.com/huggingface/lerobot/pull/983/commits/d4782cbcce35d71f209febf137061c8164e586fe)

imstevenpmwork commented on lerobot/common/datasets/video_utils.py:
```diff
@@ -252,51 +251,68 @@ def encode_video_frames(
+
+    # Reset logging level
+    if log_level is not None:
+        av.logging.restore_default_callback()
```
This might inadvertently affect other parts of the codebase if `PyAV` is used elsewhere. For now this is not the case as this is the only file using it.

CarolinePascal replied:
Maybe we could agree on a logging level for `PyAV` to be set globally for all the files using it ? I don't know if this is something we might want.

imstevenpmwork commented on lerobot/common/datasets/video_utils.py:
```diff
@@ -332,78 +348,68 @@ def __call__(self):
+        audio_info["audio.channels"] = audio_stream.channels
+        audio_info["audio.codec"] = audio_stream.codec.canonical_name
+        audio_info["audio.bit_rate"] = (
+            audio_stream.bit_rate
```
Question: Why the parentheses?

CarolinePascal replied:
pre-commit added these because of the lengthy comments. I'm re-writing them differently !

CarolinePascal replied:
Fixed in [d4782cb](https://github.com/huggingface/lerobot/pull/983/commits/d4782cbcce35d71f209febf137061c8164e586fe)

imstevenpmwork commented on lerobot/common/datasets/video_utils.py:
```diff
@@ -332,78 +348,68 @@ def __call__(self):
+        video_info["video.is_depth_map"] = False
+
+        # Calculate fps from r_frame_rate
+        video_info["video.fps"] = int(video_stream.base_rate)
```
What's the difference between these 3:
* `VideoStream.framerate`
* `VideoStream.rate`
* `Stream.base_rate`

CarolinePascal replied:
`VideoStream.rate` is just another name for `VideoStream.framerate`. `Stream.base_rate` is computed as the lowest framerate of the input, in the case where it is not perfectly constant. It is equivalent to `r_frame_rate` which was used in the previous `ffprobe` command.

aliberts commented:
I'm trying out these changes with the video benchmark using this command:
```bash
python benchmarks/video/run_video_benchmark.py \
    --output-dir outputs/video_benchmark \
    --repo-ids \
        "lerobot/xarm_push_medium_image" \
    --vcodec libx264 libx265 libsvtav1 \
    --pix-fmt yuv444p yuv420p \
    --g 2 40 None \
    --crf 5 50 None \
    --timestamps-modes 1_frame 2_frames 6_frames \
    --backends pyav \
    --num-samples 10 \
    --num-workers 5 \
    --save-frames 1
```

This runs fine on main but not on this branch:
```
Traceback (most recent call last):                                                                                                                                                                                     
  File "/Users/simon/projects/lerobot/benchmarks/video/run_video_benchmark.py", line 490, in <module>                                                                                                                  
    main(**vars(args))                                                                                                                                                                                                 
  File "/Users/simon/projects/lerobot/benchmarks/video/run_video_benchmark.py", line 366, in main                                                                                                                      
    benchmark_table += benchmark_encoding_decoding(                                                                                                                                                                    
  File "/Users/simon/projects/lerobot/benchmarks/video/run_video_benchmark.py", line 281, in benchmark_encoding_decoding                                                                                               
    benchmark_row = benchmark_decoding(
  File "/Users/simon/projects/lerobot/benchmarks/video/run_video_benchmark.py", line 219, in benchmark_decoding
    result = future.result()
  File "/opt/homebrew/Caskroom/miniconda/base/envs/lerobot/lib/python3.10/concurrent/futures/_base.py", line 451, in result
    return self.__get_result()
  File "/opt/homebrew/Caskroom/miniconda/base/envs/lerobot/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/opt/homebrew/Caskroom/miniconda/base/envs/lerobot/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/simon/projects/lerobot/benchmarks/video/run_video_benchmark.py", line 193, in process_sample
    result["mse_values"].append(mean_squared_error(original_frames_np[i], frames_np[i]))
  File "/opt/homebrew/Caskroom/miniconda/base/envs/lerobot/lib/python3.10/site-packages/skimage/metrics/simple_metrics.py", line 46, in mean_squared_error
    check_shape_equality(image0, image1)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/lerobot/lib/python3.10/site-packages/skimage/_shared/utils.py", line 627, in check_shape_equality
    raise ValueError('Input images must have the same dimensions.')
ValueError: Input images must have the same dimensions.
```
It seems that the encoded videos don't have the expected sizes.
Can you look into it?

imstevenpmwork reviewed (commented):
Left some comments. Good contribution!

CarolinePascal commented:
Side remark, I get the following error when I run the benchmark : 
```
Specified pixel format yuv444p is not supported by the libsvtav1 encoder
```
When trying with `ffmpeg` CLI this incompatibility only raises a warning : 
```
Incompatible pixel format 'yuv444p' for codec 'libsvtav1', auto-selecting format 'yuv420p'
```

CarolinePascal commented:
`hevc` works with the same warning as `libav1svt` when the pixel format is `yuv444p` ;) I added the codec check and modified the benchmark accordingly in [3c03ae3](https://github.com/huggingface/lerobot/pull/983/commits/3c03ae3629866cf2abaff66e0f7f7336f80f08c4).

I agree we should add new tests for the encoding - I would not have caught the frame size issue and the pixel format warning without the benchmark.

@imstevenpmwork I'll let you check the `av` version modification in the `pyproject.toml` to make sure it does not break anything !

CarolinePascal commented:
@aliberts it seems that most dependency managers resolve to `pyav==14.3.0` by default, and I think that your `av==12.3` is an artifact from the previous constrained install (which enforced an older version of `ffmpeg` as well).
Could you retry on your side with a fresh install to check if we still need to modify the `pyproject.toml` ?

aliberts commented:
> hevc works with the same warning as libav1svt when the pixel format is yuv444p ;) I added the codec check and modified the benchmark accordingly in [ccf2484](https://github.com/huggingface/lerobot/pull/983/commits/ccf24847bf03143429e17c0e5ffcac106fe095a7).

I added the option to select `yuv444p` color space when I wrote the video benchmark (and also because that's what we were using at the time but moving forward I'm confident that we can completely remove this parameter and simply always use `yuv420p`.

> it seems that most dependency managers resolve to pyav==14.3.0 by default

The requirement spec should not allow to install incompatible dependencies. Moreover, a dependency manager will indeed resolve to the latest version by default if you're just installing lerobot. However, if a project uses lerobot among *its* dependencies and also has e.g. `av<=14.0`, then av 14.0 will be installed. These version requirements are here to help dependency manager solve this puzzle, not to specify which exact version will be installed as this will depend from the context (and other dependencies in that context). Increased compatibility of lerobot with other tools is why we should always try to have minimum constraints on dependencies (i.e. lowest bound, as long as the code works).

So TDLR: we still need to update that pyproject ^^

CarolinePascal commented:
I set the minimal required `pyav` version to 14.2.0 in [bfe05c1](https://github.com/huggingface/lerobot/pull/983/commits/bfe05c1031acb489754422c4468446ee2c66fae1).

For now, I did not touch the pixel format. It defaults to `yuv420p` so it should not lead to major issues. I think we can remove it in a future PR, when we create new tests for the encoding for instance !

aliberts reviewed (approved):
LGTM, thank you!
